src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/linalg/FourierLinearOperator.java

```diff
@@ -0,0 +1,99 @@
+package org.broadinstitute.hellbender.tools.coveragemodel.linalg;
+
+import org.apache.commons.math3.exception.DimensionMismatchException;
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.param.ParamUtils;
+import org.jtransforms.fft.DoubleFFT_1D;
+
+import javax.annotation.Nonnull;
+import java.util.stream.IntStream;
+
+import static com.google.common.math.IntMath.isPowerOfTwo;
+import static org.broadinstitute.hellbender.utils.GATKProtectedMathUtils.nearestNeighborUniform1DInterpolate;
+import static org.broadinstitute.hellbender.utils.GATKProtectedMathUtils.smallestPowerOfTwoGreaterThan;
+
+/**
+ * A circulant linear operator specified by its eigenvalues in the Fourier space.
+ *
+ * @author Mehrtash Babadi &lt;mehrtash@broadinstitute.org&gt;
+ */
+public abstract class FourierLinearOperator<V> extends GeneralLinearOperator<V> {
+
+    protected final int dimension;
+    protected final int fftSize;
+    protected final double[] fourierFactors;
+    protected final DoubleFFT_1D fftFactory;
+
+    protected final boolean promoteToPowerOfTwo;
+
+    /**
+     * Constructs the linear operator of a given {@code dimension} for given Fourier factors
+     * Note: the length of {@code fourierFactors} must be equal to {@code floor(dimension/2) + 1}
+     *
+     * @param dimension dimension of the linear space on which the linear operator acts
+     * @param fourierFactors Fourier factors of the linear operator
+     * @param promoteToPowerOfTwo promote the dimension to the power of two
+     * @throws IllegalArgumentException if {@code fourierFactors.length != dimension/2 + 1}
+     */
+    public FourierLinearOperator(final int dimension,
+                                 @Nonnull final double[] fourierFactors,
+                                 final boolean promoteToPowerOfTwo) {
+        ParamUtils.isPositive(dimension - 1, "The dimension of the linear operator " +
+                "must be >= 2.");
+        Utils.validateArg(fourierFactors.length == dimension/2 + 1, "The length of Fourier factors must be equal" +
+                " to floor(dimension/2) + 1");
+        if (!promoteToPowerOfTwo || isPowerOfTwo(dimension)) {
+            /* dimension is already power of 2, turn off the flag */
+            this.promoteToPowerOfTwo = false;
```

Comment:
actually, `this.promoteToPowerOfTwo` is not used outside of the constructor so I ditched it altogether.